internal/repo/update.go

```diff
@@ -0,0 +1,153 @@
+package repo
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/zu1k/nali/internal/constant"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/google/go-github/v55/github"
+)
+
+var (
+	ctx      = context.Background()
+	tAsset   *github.ReleaseAsset
+	shaAsset *github.ReleaseAsset
+)
+
+func UpdateRepo() error {
+	rel, err := getLatestRelease()
+	if err != nil {
+		return fmt.Errorf("failed to get latest release: %v", err)
+	}
+
+	if !canUpdate(rel) {
+		log.Printf("current version is already the latest version, no update \n")
+		return nil
+	}
+
+	//Filtering assets by GOOS and GOARCH
+	if tAsset = getTargetAsset(rel, false); tAsset == nil {
+		return fmt.Errorf("no target asset found for %s %s", constant.OS, constant.Arch)
+	}
+	if shaAsset = getTargetAsset(rel, true); shaAsset == nil {
+		return fmt.Errorf("no sha256 asset found for %s %s", constant.OS, constant.Arch)
+	}
+
+	//Download the new version nali and its sha256
+	data, err := download(ctx, tAsset.GetID())
+	if err != nil {
+		return fmt.Errorf("failed to download asset %v: %v", tAsset.GetID(), err)
+	}
+
+	vData, err := download(ctx, shaAsset.GetID())
+	if err != nil {
+		return fmt.Errorf("failed to download asset %v: %v", tAsset.GetID(), err)
+	}
+
+	// Verifying files with sha256
+	vHash := make([]byte, sha256.Size)
+	if _, err := hex.Decode(vHash, vData[:sha256.BlockSize]); err != nil {
+		return fmt.Errorf("failed to decode sha256 hash: %v", err)
+	}
+	if !validate(data, vHash) {
+		return fmt.Errorf("failed to validate asset %v, sha256 check failed", tAsset.GetID())
+	}
+
+	// Unzip and replace nali itself
+	exe, err := os.Executable()
+	if err != nil {
+		return fmt.Errorf("could not locate executable path: %v", err)
+	}
+
+	asset, err := decompress(bytes.NewReader(data), tAsset.GetName())
+	if err != nil {
+		return fmt.Errorf("error occurred while decompress: %v", err)
+	}
+
+	log.Printf("Updating %v to %v \n", exe, rel.GetTagName())
+	if err = update(asset, exe); err != nil {
+		return fmt.Errorf("update executable failed: %v", err)
+	}
+
+	log.Printf("Successfully updated to version %v \n", rel.GetTagName())
+	return nil
+}
+
+func canUpdate(rel *github.RepositoryRelease) bool {
+	if constant.Version != "unknown version" {
+		latest, _ := semver.NewVersion(rel.GetTagName())
+		cur, _ := semver.NewVersion(constant.Version)
+
+		return latest.GreaterThan(cur)
+	}
+	return true
```

Comment:
`unknown version` means that the user compiled it manually instead of downloading it from the release, in which case we don't take the liberty of updating it to a potentially older version.

```suggestion
	return false
```